packages/duf/src/evo/data_converters/duf/importer/duf_polyface_to_evo.py

```diff
@@ -54,6 +55,49 @@ def _create_triangle_mesh_obj(name, vertices_array, indices_array, parts, epsg_c
     return triangle_mesh_go
 
 
+def indices_from_polyface(dw_facelist) -> NDArray[np.uint64]:
+    """
+    Extracts triangles from the PolyFace's FaceList.
+
+    Returns a numpy uint64 array of shape (*, 3).
+
+    The FaceList is a flat integer array. Each face is represented by 5 consecutive integers. The faces can be triangles
+    or quads.
+    - triangle: [..., 1, 2, 3, 1, -1, ...]
+    - quad: [..., 1, 2, 3, 4, -1, ...]
+    The first four integers describe the triangle/quad. The 5th integer can optionally represent colour information,
+    but we don't care about this and I don't know the details.
+
+    The FaceList is 1-indexed, and indices can be negative. Indices being negative has a special meaning for the
+    visibility of parts of the geometry, but we ignore that and force everything to be positive.
+    """
+    count = dw_facelist.Count // 5
+    assert dw_facelist.Count % 5 == 0, f"Expected a multiple of 5 indicies, but got {dw_facelist.Count}"
+
+    # The indices are 1-indexed and possibly negative
+    indices_arr = np.abs(np.fromiter(dw_facelist, dtype=np.int32, count=count * 5).reshape(count, 5)) - 1
+
+    # Triangles repeat the 0th and 3rd indices, and quads have 4 distinct indices (not considering degenerate cases)
+    quads_mask = indices_arr[:, 0] != indices_arr[:, 3]
+
+    # The first three indices are either the entire triangle we care about, or the first half of a quad
+    tris = indices_arr[:, [0, 1, 2]]
+
+    # Split off an extra triangle for each quad
+    extra_tris_from_quads = indices_arr[quads_mask][:, [2, 3, 0]]
+
+    result = np.vstack([tris, extra_tris_from_quads])
```

Comment:
[nitpick] It probably does not matter, but might we prefer the second tri for the quad to appear imediately after its first tri?  I'm unsure if there is any consistency reason we might.  Just a thought.  Its probably not worth the complication.  Could the triangle order have an impact for cache performance in certain algorithms?